src/operator/nn/dropout-inl.h

```diff
@@ -178,30 +184,17 @@ class DropoutOp {
   /*!
    * \brief Dropout kernel, compute dropout tensor
    */
-  struct DropoutKernel {
-    /*!
-     * \brief Dropout kernel function
-     * \param id Thread number (0-based representing count)
-     * \param gen Random number generator
-     * \param N Total number of items in the output
-     * \param step Step between items, related to parallelism
-     * \param dropout_out Output dropout values
-     * \param mask_out  Output mask (is multiplied to create dropout output, may be 0)
-     * \param input_data Input data to perform the dropout on
-     * \param pkeep Dropout rate (keep when the generated random number is less than this value)
-     */
+  struct BernoulliKernel {
+    /*! \brief Bernoulli kernel for generating mask */
     MSHADOW_XINLINE static void Map(int id,
                                     RandGenerator<xpu, DType> gen,
                                     const int N,
                                     const int step,
-                                    DType *dropout_out,
                                     DType *mask_out,
-                                    const DType *input_data,
                                     const real_t pkeep) {
       RNG_KERNEL_LOOP(xpu, DType, id, gen, N, step, {
         const real_t rand_num = static_cast<real_t>(genImpl.uniform());
         mask_out[i] = mshadow_op::threshold::Map<real_t>(rand_num, pkeep) * (1.0f / pkeep);
-        dropout_out[i] = input_data[i] * mask_out[i];
```

Comment:
I am not saying it needs to be done, but have you considered merging this operation with the other kernel, perhaps by deriving from broadcast_kernel or passing a modified version of the mul OP to broadcast_kernel?
Making two full passes across the memory is going to cause a performance hit due to both OMP overhead as well as CPU cache.

Reply:
I am not sure I understand you clearly. 
I separate the original dropout kernel into two parts: 1) BernoulliKernel 2) broad_cast Mul

Reply:
I am not sure I understand you clearly. 
I separate the original dropout kernel into two parts: 1) BernoulliKernel 2) broad_cast mul 
so that we can enable axes support for variational dropout.

Thx

Reply:
Right. What's the performance impact of using two kernels instead of one?

Reply:
The mask size is different with input size.

For example: input can be 10x5x2x3, the axes are (1,2)
then the mask will be the size of 10x1x1x3